### PR TITLE
Rnd211 32

### PR DESCRIPTION
**Pull request by Seweryn-Zielas-Mobica**
Add working  ar9285_attach and its dependencies.

### CODE REVIEW

File: sys/dev/athn/if_athn_usb.c [hunk 1]
```diff
@@ -603,16 +615,104 @@ athn_usb_verify_fw_ver(struct athn_softc *sc, struct ar_wmi_fw_version *img_ver)
...
+	bf = _athn_getbuf(sc);
+	return (bf);
+}
+
+static void
+athn_freebuf(struct athn_softc *sc, struct athn_usb_data *bf)
+{
+	struct athn_usb_softc *usc = sc->sc_usc;
+
+	ATHN_ASSERT_LOCKED(usc);
```
Comment (Mikolaj-Filar-Mobica) [inline sys/dev/athn/if_athn_usb.c]: _sc_ shall be passed

File: sys/dev/athn/if_athn_usb.c [hunk 2]
```diff
@@ -441,7 +453,7 @@ athn_usb_attach(device_t self)
 	sc->ops.write = athn_usb_write;
 	sc->ops.write_barrier = athn_usb_write_barrier;
 
-	mtx_init(&sc->sc_mtx, device_get_nameunit(self), MTX_NETWORK_LOCK, MTX_DEF);
+	mtx_init(&sc->sc_mtx, device_get_nameunit(self), MTX_NETWORK_LOCK, MTX_DEF | MTX_RECURSE);
```
Comment (Mikolaj-Filar-Mobica) [inline sys/dev/athn/if_athn_usb.c]: Are you sure we want to go with recurse?

**Review by AlicjaMobica — APPROVED** [review]
It's looking good

File: sys/dev/athn/headers/athnvar.h [hunk 3]
```diff
@@ -455,7 +459,117 @@ struct athn_ops {
...
+	    calTargetPower2GHT20[AR9285_NUM_2G_20_TARGET_POWERS];
+	struct	ar_cal_target_power_ht
+	    calTargetPower2GHT40[AR9285_NUM_2G_40_TARGET_POWERS];
+	uint8_t	ctlIndex[AR9285_NUM_CTLS];
+	struct	ar9285_cal_ctl_data ctlData[AR9285_NUM_CTLS];
+	uint8_t	padding;
+} __packed;
+
 struct athn_softc {
+	struct athn_usb_softc *sc_usc;
```
Comment (Mikolaj-Filar-Mobica) [inline sys/dev/athn/headers/athnvar.h:572]: Are you sure it's needed?

File: sys/dev/athn/headers/athnvar.h [hunk 4]
```diff
@@ -554,6 +670,7 @@ struct athn_softc {
 	struct athn_txq			txq[31];
 
 	struct mtx		sc_mtx;
+	struct sx 		sc_sx;
```
Comment (Mikolaj-Filar-Mobica) [inline sys/dev/athn/headers/athnvar.h]: Seems to be not needed.

File: sys/dev/athn/headers/if_athn_usb.h [hunk 5]
```diff
@@ -428,6 +428,8 @@ struct athn_usb_data {
 	struct usb_xfer		*xfer;
 	char *buf;
 	usb_frlength_t buflen;
+	struct mbuf *m;
+	struct ieee80211_node	*ni;
```
Comment (Mikolaj-Filar-Mobica) [inline sys/dev/athn/headers/if_athn_usb.h:432]: Are you sure we need _ni_ in this structure?

File: sys/dev/athn/if_athn_usb.c [hunk 6]
```diff
@@ -424,6 +427,15 @@ athn_usb_attach(device_t self)
 	usb_error_t error;
 	struct usb_endpoint *ep, *ep_end;
 
+	{
```
Comment (Mikolaj-Filar-Mobica) [inline sys/dev/athn/if_athn_usb.c]: Do we need this affinity? I can't see it e.g. in otus

File: sys/dev/athn/if_athn_usb.c [hunk 7]
```diff
@@ -623,6 +723,8 @@ athn_usb_attachhook(device_t self)
 		printf("Could not load firmware\n");
 		return;
 	}
+	DELAY(1000 *1000);
```
Comment (Mikolaj-Filar-Mobica) [inline sys/dev/athn/if_athn_usb.c:747]: Do we want to have this delay?

File: sys/dev/athn/if_athn_usb.c [hunk 8]
```diff
@@ -1207,17 +1332,18 @@ athn_if_bulk_tx_callback(struct usb_xfer *xfer, usb_error_t error)
...
 	case USB_ST_TRANSFERRED:
 		data = STAILQ_FIRST(&usc->sc_tx_active[which]);
 		if (data == NULL)
 			goto tr_setup;
-		device_printf(sc->sc_dev, "%s: transfer done %p\n", __func__, data);
+		// device_printf(sc->sc_dev, "%s: transfer done %p\n", __func__, data);
 		STAILQ_REMOVE_HEAD(&usc->sc_tx_active[which], next);
 		athn_usb_txeof(xfer, data);
+		athn_freebuf(sc, data);
 		STAILQ_INSERT_TAIL(&usc->sc_tx_inactive, data, next);
```
Comment (Mikolaj-Filar-Mobica) [inline sys/dev/athn/if_athn_usb.c:1370]: You didn't edit that, but I wonder where that line came from.

File: sys/dev/athn/if_athn_usb.c [hunk 9]
```diff
@@ -1588,12 +1717,17 @@ athn_usb_read_rom(struct athn_softc *sc)
...
 	addr = AR_EEPROM_OFFSET(sc->eep_base);
 	for (i = 0; i < sc->eep_size / 16; i++) {
+		DELAY(3000);
 		for (j = 0; j < 8; j++, addr += 4)
 			addrs[j] = htobe32(addr);
 		error = athn_usb_wmi_xcmd(usc, AR_WMI_CMD_REG_READ,
 		    addrs, sizeof(addrs), vals);
+		DELAY(3000);
 		if (error != 0)
+		{
```
Comment (Mikolaj-Filar-Mobica) [inline sys/dev/athn/if_athn_usb.c:1750]: Code formatting. '{' shall be in the line above.

File: sys/dev/athn/if_athn_usb.c (hunk 1 above)
Comment (Seweryn-Zielas-Mobica) [reply sys/dev/athn/if_athn_usb.c]: Fixed

File: sys/dev/athn/if_athn_usb.c (hunk 2 above)
Comment (Seweryn-Zielas-Mobica) [reply sys/dev/athn/if_athn_usb.c]: Removed

File: sys/dev/athn/headers/athnvar.h (hunk 3 above)
Comment (Seweryn-Zielas-Mobica) [reply sys/dev/athn/headers/athnvar.h:572]: For now it is, I used it in some new functions related to raw_xmit. Later we can change parameters in that functions if we decide to change approach with them.

File: sys/dev/athn/headers/athnvar.h (hunk 4 above)
Comment (Seweryn-Zielas-Mobica) [reply sys/dev/athn/headers/athnvar.h]: Forgot to remove it.

File: sys/dev/athn/headers/if_athn_usb.h (hunk 5 above)
Comment (Seweryn-Zielas-Mobica) [reply sys/dev/athn/headers/if_athn_usb.h:432]: It is needed in athn_usb_txeof.

File: sys/dev/athn/if_athn_usb.c (hunk 6 above)
Comment (Seweryn-Zielas-Mobica) [reply sys/dev/athn/if_athn_usb.c]: I think we don't. I added them for experiment to check if something changes. This should prevent change of the context/thread after sleeping but I'm not sure it this works properly.

File: sys/dev/athn/if_athn_usb.c (hunk 7 above)
Comment (Seweryn-Zielas-Mobica) [reply sys/dev/athn/if_athn_usb.c:747]: I would leave it. Same mechanism is present in otus after loading firmware to device.

File: sys/dev/athn/if_athn_usb.c (hunk 8 above)
Comment (Seweryn-Zielas-Mobica) [reply sys/dev/athn/if_athn_usb.c:1370]: I added it with raw_xmit.

File: sys/dev/athn/if_athn_usb.c (hunk 9 above)
Comment (Seweryn-Zielas-Mobica) [reply sys/dev/athn/if_athn_usb.c:1750]: I would leave it for now. We should use formatting tool at the end of implementing driver for all changes that we made.

File: sys/dev/athn/headers/athnvar.h (hunk 3 above)
Comment (Mikolaj-Filar-Mobica) [reply sys/dev/athn/headers/athnvar.h:572]: ~~Now we get the variable with: `struct athn_usb_softc *usc = device_get_softc(self);` I wonder if we will have two identical variables with different values. ̶~~
Agree with this change. It's the first element in the `sc` structure, so it should work fine.

File: sys/dev/athn/if_athn_usb.c (hunk 7 above)
Comment (Mikolaj-Filar-Mobica) [reply sys/dev/athn/if_athn_usb.c:747]: We already do it in `athn_usb_transfer_firmware`. After FW upload we send message and wait up to 1 second for a response:
```C
	usc->wait_msg_id = AR_HTC_MSG_READY;
	error = usbd_do_request(usc->sc_udev, &usc->sc_sc.sc_mtx, &req, NULL);	

	/* Wait at most 1 second for firmware to boot. */
	device_printf(usc->sc_sc.sc_dev, "%s: Waiting for firmware to boot\n", __func__);
	if (error == 0 && usc->wait_msg_id != 0)
		error = msleep(&usc->wait_msg_id, &usc->sc_sc.sc_mtx, PCATCH, "athnfw", hz);

	if (usc->wait_msg_id == 0) {
		device_printf(usc->sc_sc.sc_dev, "%s: Firmware booted successfully!\n", __func__);
	}
```
Anyway, I'm happy to keep it.

File: sys/dev/athn/if_athn_usb.c (hunk 8 above)
Comment (Mikolaj-Filar-Mobica) [reply sys/dev/athn/if_athn_usb.c:1370]: I mean the line below your change but nvm.

File: sys/dev/athn/athn.c [hunk 10]
```diff
@@ -2574,15 +2576,15 @@ int
...
 	uint32_t reg, def_ant, sta_id1, cfg_led, tsflo, tsfhi;
 	int i, error;
 
 	/* XXX not if already awake */
 	if ((error = athn_set_power_awake(sc)) != 0) {
-		// TOTO missing field 'dv_xname' in 'struct device'
-		// printf("%s: could not wakeup chip\n", sc->sc_dev.dv_xname);
+		device_printf(sc->sc_dev, "%s: ould not wakeup chip\n", __func__);
+		return (error);
 		return (error);
```
Comment (Mikolaj-Filar-Mobica) [inline sys/dev/athn/athn.c:2587]: Duplicated `return (error);` line

**Review by Mikolaj-Filar-Mobica — APPROVED** [review]

File: sys/dev/athn/athn.c (hunk 10 above)
Comment (Seweryn-Zielas-Mobica) [reply sys/dev/athn/athn.c:2587]: Removed

**Pull request merged**